File: pipeline/processing/app/schemas/dataset.py
"""Request and response models for the dataset and prediction surfaces."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class DatasetSource(BaseModel):
    """What the project's scientific tables currently hold.

    Separated from the dataset itself so the client can tell an empty project apart from an
    unimplemented builder — the first is fixed by extracting more papers, the second is not
    fixable by the user at all.
    """

    experiments: int = 0
    ingredients: int = 0
    indicators: int = 0
    measurements: int = 0
    #: Indicators carrying a threshold. Without one, no survival label can be derived, so
    #: this is the count that decides whether a build could produce anything.
    indicators_with_threshold: int = 0


class DatasetPreview(BaseModel):
    """The project's flat, prediction-ready dataset — or the reason there isn't one."""

    project_id: int
    status: Literal["not_built", "not_implemented", "ready"] = "not_built"
    columns: list[str] = Field(default_factory=list)
    row_count: int = 0
    #: A head of the data, never the whole thing. The full dataset goes to the prediction
    #: service directly; this is for the screen.
    rows: list[dict[str, Any]] = Field(default_factory=list)
    source: DatasetSource = Field(default_factory=DatasetSource)


class JobAccepted(BaseModel):
    """The only thing a "start work" endpoint returns.

    Progress is followed through `GET /api/jobs/{job_id}/events`, the same stream the
    extraction workflows use; there is no per-workflow status endpoint to poll.
    """

    job_id: int
    status: str

File: pipeline/frontend/src/api/dataset.ts
/**
 * The processed dataset and the prediction surface, both served by the processing service.
 *
 *   useDataset       -> GET  /projects/{id}/dataset
 *   useBuildDataset  -> POST /projects/{id}/dataset/build   -> { job_id }
 *   usePredictionStatus -> GET /projects/{id}/prediction
 *
 * The builder behind these is not written yet: a build reports what it would have consumed
 * and produces no rows. The wiring is real regardless — the screens render the empty
 * result, so what remains is the builder, not the plumbing around it.
 */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { get, post } from './client'
import { keys } from './keys'
import type { JobAccepted } from '../types/workspace'

const base = (projectId: number) => `/projects/${projectId}`

/** What the project's scientific tables currently hold. */
export interface DatasetSource {
  experiments: number
  ingredients: number
  indicators: number
  measurements: number
  /** Indicators carrying a threshold — without one, nothing can be labelled. */
  indicators_with_threshold: number
}

export interface DatasetPreview {
  project_id: number
  status: 'not_built' | 'not_implemented' | 'ready'
  columns: string[]
  row_count: number
  rows: Record<string, unknown>[]
  source: DatasetSource
}

export interface PredictionStatus {
  project_id: number
  available: boolean
  reason: string
  engines: string[]
  source: DatasetSource
}

export function useDataset(projectId: number) {
  return useQuery({
    queryKey: keys.dataset.preview(projectId),
    queryFn: () => get<DatasetPreview>(`${base(projectId)}/dataset`),
    enabled: Number.isFinite(projectId),
  })
}

export function useBuildDataset(projectId: number) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: () => post<JobAccepted>(`${base(projectId)}/dataset/build`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: keys.jobs.all() })
    },
  })
}

export function usePredictionStatus(projectId: number) {
  return useQuery({
    queryKey: keys.prediction.status(projectId),
    queryFn: () => get<PredictionStatus>(`${base(projectId)}/prediction`),
    enabled: Number.isFinite(projectId),
  })
}
